spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java

```diff
@@ -122,11 +122,7 @@ public static Expression convert(Filter filter) {
 
         case IN:
           In inFilter = (In) filter;
-          Expression in = alwaysFalse();
-          for (Object value : inFilter.values()) {
-            in = or(in, equal(inFilter.attribute(), convertLiteral(value)));
-          }
-          return in;
+          return in(inFilter.attribute(), inFilter.values());
```

Comment:
What if values contain null? Can we add a test for this?

Reply:
The values cannot contain null. It is not allowed when you [create the `in` expression](https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/Expressions.java#L191). There's a secondary check when [creating a literal](https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/Literals.java#L59) as well.

Reply:
Yes, but what if Spark passes a list of values that includes null?

Reply:
Now that I read this a second time, I think you may be talking about when Spark passes a null value here?

I agree that's a problem. This should detect and filter out null values because Iceberg's `in` expression will throw an exception if it encounters them. If there is a null, then this can return `or(isNull(attr), in(attr, nonNullValues))`

Reply:
Thanks @aokolnychyi for the comments. I will add additional tests for those cases. 
@rdblue I am thinking if we can add this transformation `(in -> or(isNull, in))` into iceberg's `Expressions` so each caller does not need to repeatedly implement this logic.

Reply:
Actually, null is never equal to null.
```
null IN (1, 2, 3) -> null
null IN (1, 2, 3, null) -> null
1 IN (1, 2, 3) -> true
1 IN (1, 2, 3, null) -> true
-1 IN (1, 2, 3) -> false
-1 IN (1, 2, 3, null) -> null
```
I wonder if we can simply filter out null values to avoid the exception.

Reply:
@aokolnychyi is right. `null` is never equal to `null`.

@jun-he, that's the reason why we shouldn't automatically transform in `Expressions`. By not allowing callers to create predicates like `in(1, null, 2)` we avoid the problem in most Iceberg.

We still need to fix the case where Spark passes in `null`, though. I think that Anton is right and we can simply filter `null` out of the list. A null in the list will never cause a value to be accepted. It will only cause the filter to return `null` instead of `false`, which is handled like `false` when filtering: if the filter evaluates to `null` then the row is not selected.